Partial payment test

forestb opened the pull request:
Added an additional test where LevelUp was the second payment and the discount amount available covered the balance in full.

mmercurio commented on LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs:
```diff
@@ -297,5 +297,74 @@ public void SplitTenderExample_LevelUp_Then_Cash_With_Adjusted_Exemption_And_Adj
+            int availableDiscountAmount = 1000;
+
+            // POS applies $10.00 pretax discount to check and updates subtotal and tax
+            currentTax = (originalSubtotal - availableDiscountAmount) / TAX_RATE;
```
Why use `TAX_RATE` here to calculate tax, but not above?  I realize the end result is the same, but the style difference is odd and may lead one to wonder if there's a problem.

forestb replied:
This should be addressed by the new `Check` class, but it was a good catch.

forestb commented:
@mmercurio, there are other scenarios we should add to test cases.  I'll write up a ticket to add those.

forestb commented on LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs:
```diff
@@ -106,16 +95,21 @@ public void SplitTenderExample_Cash_Then_LevelUp()
             // Updated check:
             //   PAID cash:     $10
             //   TOTAL DUE:     $12
+            check.ApplyTender(1000);
```
The `Check` class supports applying tenders.

forestb commented on LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs:
```diff
@@ -151,26 +132,22 @@ public void SplitTenderExample_Cash_Then_LevelUp()
-            totalTaxAmount = itemsSubtotalAmount / TAX_RATE;
-            totalOutstandingAmount = itemsSubtotalAmount + totalTaxAmount - paidCashAmount;
-            spendAmount = totalOutstandingAmount + availableDiscountAmount;
+            check.ApplyDiscount(availableDiscountAmount);
```
The `Check` class supports applying discounts.

forestb commented:
@mmercurio, I've added a `Check` class to streamline some of the calculations and help make writing additional tests a bit more intuitive.  I've also updated the pre-existing tests to use the new class.  We were essentially simulating a POS already, and this just streamlines the logic.  Let me know what you think.

forestb commented:
Also added [Jira ticket POS-363](https://levelup.atlassian.net/browse/POS-363) for adding some testing to edge cases where discount is applied and exemption amounts.

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs:
```diff
@@ -0,0 +1,60 @@
+    /// </summary>
+    internal class Check
+    {
+        #region Private data members
```
lose the region

forestb replied:
Lost it.

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs:
```diff
@@ -0,0 +1,60 @@
+        #endregion
+
+        /// <summary>
+        /// The current amount, in cents, owed by a customer, in cents, at any given moment.
```
what's with the echo?

forestb replied:
Roger, Roger.

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs:
```diff
@@ -0,0 +1,60 @@
+        public int TotalOutstandingAmount => AdjustedSubtotal + AdjustedTax - TotalTenders;
+
+        /// <summary>
+        /// The total amount of tax due, in cents, on a check, in cents, at any given moment.
```
same deal

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs:
```diff
@@ -0,0 +1,60 @@
+        /// By passing in a total amount owed and the tax due on a check, we can simulate a simplistic point of sale
+        /// for testing purposes.
+        /// </summary>
+        /// <param name="total">The intial amount due on the check, in cents.</param>
```
including tax, correct?

forestb replied:
Yes, added to description.

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/Helpers/Check.cs:
```diff
@@ -0,0 +1,60 @@
+
+        private int StartingTotal { get; }
+        private int StartingTax { get; }
+        private int StartingSubtotal => StartingTotal - StartingTax;
```
Subtotal --> UntaxedTotal, maybe?

forestb replied:
I think the common language in the restaurant world is subtotal.  How strongly do you feel about this?  Is there a reason "Subtotal" is not correct?

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/LevelUp.Pos.ProposedOrders.Tests.csproj:
```diff
@@ -48,6 +48,7 @@
   <ItemGroup>
     <Compile Include="CalculatorTests.cs" />
     <Compile Include="Data\CalculatorTestData.cs" />
+    <Compile Include="Helpers\Check.cs" />
```
Helpers/ --> Mocks/ ?  Is this technically a mock?

forestb replied:
Good call.

devlaf commented on LevelUp.Pos.ProposedOrders.Tests/SplitTenderTests.cs:
```diff
@@ -287,8 +248,69 @@ public void SplitTenderExample_LevelUp_Then_Cash_With_Adjusted_Exemption_And_Adj
+            // Create completed order: expected values
+            AdjustedCheckValues expectedCompletedOrderValues =
+                new AdjustedCheckValues(
+                    spendAmount: 1100,      // spendAmount will include appliedDiscount and remaining tax
```
I spent a bunch of time trying to figure out why this number is what it is.  Maybe worth adding a comment.

forestb replied:
✔️

devlaf commented:
Well I'm happy with the changes.  LGTM
